File: javascript/pig-latin/pig-latin.js
const regex = /^[aeiou]|^xr|^y[^aeiou]/;

export const translate = (phrase) => 
    phrase.split(' ').map(translateWord).join(' ');

export const translateWord = (word) => {
    if (regex.test(word)){
        return word + 'ay';
    }

    let newWord = word.startsWith('qu') 
                    ? word.substring(2) + 'qu' : word.substring(1) + word[0];

    return translateWord(newWord);
};

File: csharp/pig-latin/PigLatin.cs
using System;
using System.Linq;
using System.Text.RegularExpressions;

public static class PigLatin
{
    private static string regex = "(^[aeiou]|^xr|^y[^aeiou]).*";

    public static string Translate(string phrase) =>
        string.Join(" ", phrase.Split(" ").Select(TranslateWord));

    public static string TranslateWord(string word) 
    {
        if (Regex.IsMatch(word, regex)) {
            return word + "ay";
        }

        string newWord = Regex.IsMatch(word, "^qu.*")
                ? word.Substring(2) + "qu" 
                : word.Substring(1) + word.Substring(0, 1);

        return TranslateWord(newWord);
    }
}
